ibis/expr/operations.py

```diff
@@ -12,25 +12,228 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import operator
-import six
-import itertools
 
+import six
 import toolz
+import operator
+import itertools
+import collections
 
-from ibis.expr.types import TableColumn  # noqa
-
+from functools import partial
 from ibis.expr.schema import HasSchema, Schema
-from ibis.expr.rules import value, string, number, integer, boolean, list_of
-from ibis.expr.types import (Node, as_value_expr, Expr,
-                             ValueExpr, ColumnExpr, TableExpr,
-                             ValueOp, _safe_repr, distinct_roots)
-import ibis.common as com
-import ibis.expr.datatypes as dt
-import ibis.expr.rules as rules
-import ibis.expr.types as ir
+
 import ibis.util as util
+import ibis.common as com
 import ibis.compat as compat
+import ibis.expr.types as ir
+import ibis.expr.rules as rlz
+import ibis.expr.datatypes as dt
+
+from collections import OrderedDict
+
+
+def _safe_repr(x, memo=None):
+    return x._repr(memo=memo) if isinstance(x, (ir.Expr, Node)) else repr(x)
+
+
+# TODO: move to analysis
+def distinct_roots(*expressions):
+    roots = toolz.concat(
+        expression._root_tables() for expression in expressions
+    )
+    return list(toolz.unique(roots, key=id))
+
+
+class TypeSignature(OrderedDict):
+
+    __slots__ = tuple()
+
+    def __call__(self, *args, **kwargs):
+        if len(args) > len(self.keys()):
+            raise TypeError('takes {} positional arguments but {} were '
+                            'given'.format(len(args), len(self.keys())))
+
+        result = []
+        for i, (name, rule) in enumerate(self.items()):
+            if i < len(args):
+                if name in kwargs:
+                    raise TypeError('got multiple values for argument'
+                                    '{}'.format(name))
+                value = rule.call(args[i])
+            elif name in kwargs:
+                value = rule.call(kwargs[name])
+            else:
+                value = rule.call()
+
+            result.append((name, value))
+
+        return result
+
+    def names(self):
+        return self.keys()
+
+
+class OperationMeta(type):
+
+    @classmethod
+    def __prepare__(metacls, name, bases, **kwds):
```

Comment:
This doesn't exist on Python 2.7, so we'll need to find another way of writing this class. We may have a module that we import it from depending on the version of python so that we can continue to use `__prepare__` in Python >=3